source/common/http/http1/codec_impl.cc

```diff
@@ -421,12 +421,11 @@ void ConnectionImpl::onHeaderValue(const char* data, size_t length) {
     // Ignore trailers.
     return;
   }
-  // http-parser should filter for this
-  // (https://tools.ietf.org/html/rfc7230#section-3.2.6), but it doesn't today. HeaderStrings
-  // have an invariant that they must not contain embedded zero characters
-  // (NUL, ASCII 0x0).
+  // http-parser > 2.9.0 does filter for this
+  // (https://tools.ietf.org/html/rfc7230#section-3.2.6). HeaderStrings have an invariant that
+  // they must not contain embedded zero characters (NUL, ASCII 0x0).
   if (absl::string_view(data, length).find('\0') != absl::string_view::npos) {
-    throw CodecProtocolException("http/1.1 protocol error: header value contains NUL");
+    throw CodecProtocolException("http/1.1 protocol error: HPE_INVALID_HEADER_TOKEN");
```

Comment:
Why change the message? I think we don't expect this to happen anymore and it could be made an `ASSERT`, but this is still super scary and has burned us once, so I would leave it there.

Reply:
Without this change the asan and tsan builds would fail with the error:
```
test/common/http/http1/codec_impl_test.cc:305: Failure
Value of: std::string(e.what())
Expected: is equal to 0x256508 pointing to "http/1.1 protocol error: header value contains NUL"
  Actual: "http/1.1 protocol error: HPE_INVALID_HEADER_TOKEN"
```
What I _think_ is happening is that the test was expecting the code in `source/common/http/http1/codec_impl.cc` to generate the error with a header value containing NUL. But with the change to http-parser in at https://github.com/nodejs/http-parser/blob/v2.9.2/http_parser.c#L1509-L1512 http-parser is catching the error and returning a different value.

Considering the sensitivity of this code re: the CVE I suggest I back out the http-parser upgrade and create an issue for it to be studied in more depth.